### PR TITLE
third_party/utf8_range: support arm neon

### PR DESCRIPTION
**Pull request by cyb70289**
Protobuf uses utf8_range library for utf8 string validation.
Currently, only SSE implementation is integrated.
This patch adapts utf8_range Neon implementation to protobuf.

### CODE REVIEW

**Comment by cyb70289** [comment]
hi, I'm author of utf8_range, glad to see my lib adopted by protobuf.
This patch adapts utf8_range Neon implementation to protobuf. Please review. Thanks.

**Comment by cyb70289** [comment]
@acozzette will you have a look at this pr? or someone else can help?

**Comment by tonyliaoss** [comment]
Hello Yibo,

I'll reassign this to @danlark1 who is our ARM SIMD expert. Thanks for making this contribution.

In the meantime, I'll approve this for integration testing.

**Review by tonyliaoss — APPROVED** [review]
Approve for integration testing.

(Do not submit until another approval from @danlark1)

**Comment by cyb70289** [comment]
Hmm..., looks this pr leads to java and ruby linux aarch64 job failure, will check.

**Comment by cyb70289** [comment]
Should have fixed ruby error. Please help start CI jobs.
Not sure of java failure. Does java use this c++ utf8 validation at all?

File: third_party/utf8_range/utf8_range.c
Comment (danlark1) [inline third_party/utf8_range/utf8_range.c:140]: I think `__aarch64__` is also required as 	vqtbl1q_u8 is only for A64 

https://developer.arm.com/architectures/instruction-sets/intrinsics/vqtbl1q_u8

File: third_party/utf8_range/utf8_range_neon.inc
Comment (danlark1) [inline third_party/utf8_range/utf8_range_neon.inc:4]: Internally when submitting x86 version we changed the comments so that they are more understandable to several people who reviewed the code. I think having a slightly different set of comments in another file might be confusing. I suggest deleting comments here and refer to the sse version of file

It does not mean your comments are bad or anything wrong with them, thank you for doing the work anyway! It's more for the clarity of the future reader when we intend to change anything

**Review by danlark1 — COMMENTED** [review]
Thanks!

File: third_party/utf8_range/utf8_range_neon.inc
Comment (cyb70289) [reply third_party/utf8_range/utf8_range_neon.inc:4]: Indeed. I thought about leaving this file without comment and only reference the sse code.
One catch is the adjustment code is different from sse (neon table lookup is more convenient in this case).
I will delete similar comment and leave only the adjustment related.

File: third_party/utf8_range/utf8_range.c
Comment (cyb70289) [reply third_party/utf8_range/utf8_range.c:140]: Good catch!

I try to use only macros defined by ACLE [1].
From the text below, the updated code checks both __ARM_NEON and __ARM_64BIT_STATE.
These two macros are defined since gcc-5.4 and clang-9.0. [2]
[1] https://arm-software.github.io/acle/main/acle.html#availability-of-advanced-simd-intrinsics
[2] https://godbolt.org/z/s918cP8fa

```
14.2.1 Availability of Advanced SIMD intrinsics
Advanced SIMD support is available if the __ARM_NEON macro is predefined (see Advanced SIMD architecture extension (Neon)). In order to access the Advanced SIMD intrinsics, it is necessary to include the <arm_neon.h> header.

  #if __ARM_NEON
  #include <arm_neon.h>
    /* Advanced SIMD intrinsics are now available to use.  */
  #endif
Some intrinsics are only available when compiling for the AArch64 execution state. This can be determined using the __ARM_64BIT_STATE predefined macro (see Instruction set architecture (A32/T32/A64)).
```

**Comment by cyb70289** [comment]
A bit struggling about the jave linux aarch64 job failure. Looks it's related to this PR. Any suggestion is welcomed.

**Comment by cyb70289** [comment]
Hopefully all issues are fixed. Please help trigger CI.

**Comment by danlark1** [comment]
Approval for code. Thank you a lot!

**Comment by cyb70289** [comment]
How to check CI error ```feedback/copybara - google internal checks FAILED```?

**Comment by tonyliaoss** [comment]
Hi Yibo --

This error message about Copybara failures is saying that it's pending internal integration tests to be run before we can pull it into Google. This CL looks good so far, we just need to get approvals internally to get it integrated into our monorepo (and then we can close this PR).

There is no action needed on your part. We've been a bit busy these few days but hopefully we can get this merged soon.

**Comment by tonyliaoss** [comment]
This is failing our internal integration tests.

I haven't fully debugged what's going on, but I can tell a behavior change happened in the SSE (non-neon) codepath, due to the changes that you [mentioned](https://github.com/protocolbuffers/protobuf/pull/18126#issuecomment-2362827138) in a previous comment, in these two places:

Screenshot 2024-09-27 at 9 12 47 PM

Screenshot 2024-09-27 at 9 12 59 PM

If I revert these two changes, the regression disappears.

Specifically it seems like the change on this line is somewhat problematic:

```
   /* Find previous codepoint (not 80~BF) */
   data -= utf8_range_CodepointSkipBackwards(_mm_extract_epi32(prev_input, 3));
   if (return_position) {
-    return (data - (end - len)) +
+    return (data - data_original) +
            utf8_range_ValidateUTF8Naive(data, end, return_position);
   }
   /* Test if there was any error */
```

`end - len` is not always equal to `data_original`. I can't quite figure out why it might be unequal though.

**Comment by tonyliaoss** [comment]
Oh I think I know what the problem is. `data` might be skipped forward due to line 182: https://github.com/protocolbuffers/protobuf/pull/18126/files#diff-4f84906404b1aa9c995fb03b21950c498c4a4b86381887686ec7c7de66fb9834L182

```
static FORCE_INLINE_ATTR inline size_t utf8_range_Validate(
    const char* data, size_t len, int return_position) {
  if (len == 0) return 1 - return_position;
  const char* const end = data + len;      //// <---- END IS SET HERE
  data = utf8_range_SkipAscii(data, end);  //// <---- DATA IS SKIPPED FORWARD
  /* SIMD algorithm always outperforms the naive version for any data of
     length >=16.
   */
  if (end - data < 16) {
    return (return_position ? (data - (end - len)) : 0) +
           utf8_range_ValidateUTF8Naive(data, end, return_position);
  }
#if defined(__SSE4_1__) || (defined(__ARM_NEON) && defined(__ARM_64BIT_STATE))
  return utf8_range_ValidateUTF8Simd(data, end, return_position);
#else
  return (return_position ? (data - (end - len)) : 0) +
         utf8_range_ValidateUTF8Naive(data, end, return_position);
#endif
}
```

If I assign `const char* const data_original = data;` in the first line of `utf8_range_Validate`, everything works as intended.

If, instead, `data_original` is assigned within `utf8_range_ValidateUTF8Simd`, and if data is skipped forward due to `utf8_range_SkipAscii(data, end)`, then the following statement

```
data_original == end - len;
```

is not true.

**Comment by tonyliaoss** [comment]
I'm going to rerun tests with the following patchset and see what happens:

##### `utf8_range.c`

```
@@ -178,19 +178,22 @@
 static FORCE_INLINE_ATTR inline size_t utf8_range_Validate(
     const char* data, size_t len, int return_position) {
   if (len == 0) return 1 - return_position;
+  // Save buffer start address for later use
+  const char* const data_original = data;
   const char* const end = data + len;
   data = utf8_range_SkipAscii(data, end);
   /* SIMD algorithm always outperforms the naive version for any data of
      length >=16.
    */
   if (end - data < 16) {
-    return (return_position ? (data - (end - len)) : 0) +
+    return (return_position ? (data - data_original) : 0) +
            utf8_range_ValidateUTF8Naive(data, end, return_position);
   }
 #if defined(__SSE4_1__) || (defined(__ARM_NEON) && defined(__ARM_64BIT_STATE))
-  return utf8_range_ValidateUTF8Simd(data, end, return_position);
+  return utf8_range_ValidateUTF8Simd(
+      data_original, data, end, return_position);
 #else
-  return (return_position ? (data - (end - len)) : 0) +
+  return (return_position ? (data - data_original) : 0) +
          utf8_range_ValidateUTF8Naive(data, end, return_position);
 #endif
 }
```

##### `utf8_range_neon.inc`

```
@@ -7,7 +7,8 @@
  */
 
 static FORCE_INLINE_ATTR inline size_t utf8_range_ValidateUTF8Simd(
-    const char* data, const char* end, int return_position) {
+    const char* const data_original, const char* data,
+    const char* end, int return_position) {
   const uint8x16_t first_len_tbl = {
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 3,
   };
@@ -57,7 +58,6 @@
   uint8x16_t prev_first_len = vdupq_n_u8(0);
   uint8x16_t error = vdupq_n_u8(0);
 
-  const char* const data_original = data;
   while (end - data >= 16) {
     const uint8x16_t input = vld1q_u8((const uint8_t*)data);
 ```

##### `utf8_range_sse.inc`

```
@@ -3,7 +3,8 @@
 #include <tmmintrin.h>
 
 static FORCE_INLINE_ATTR inline size_t utf8_range_ValidateUTF8Simd(
-    const char* data, const char* end, int return_position) {
+    const char* const data_original, const char* data,
+    const char* end, int return_position) {
   /* This code checks that utf-8 ranges are structurally valid 16 bytes at once
    * using superscalar instructions.
    * The mapping between ranges of codepoint and their corresponding utf-8
@@ -154,8 +155,6 @@
   __m128i prev_first_len = _mm_set1_epi8(0);
   __m128i error = _mm_set1_epi8(0);
 
-  // Save buffer start address for later use
-  const char* const data_original = data;
   while (end - data >= 16) {
     const __m128i input = _mm_loadu_si128((const __m128i*)(data));
```

**Comment by cyb70289** [comment]
Ouch! It's very lucky that internal test catches this bug.😓
Thank you for the debugging.

**Pull request closed without merging**
